test_python/test_tc.py

```diff
@@ -12,89 +12,499 @@
+        cache_filename = "/tmp/" + str(uuid.uuid4())
+        convolution = tc.define_with_autograd(
+            conv,
+            forward_entry_points=("convolution", ),
+            backward_entry_points=("convolution_grad", ),
+            cache_filename=cache_filename,
+            tuner_config=tuner_config)
+        I = Variable(torch.randn(N, C, H, W).cuda(), requires_grad=True)
+        W = Parameter(torch.randn(O, C, kH, kW).cuda())
+        # First occurrence on a new cache filename triggers tuning
+        out, = convolution(I, W)
+        out.sum().backward()
+        # Subsequent occurrences do not
+        out, = convolution(I, W)
+        out.sum().backward()
+        # TODO: test results!!!
 
 
-class TensorDotTest(TestCase):
-    def build_tc_problem(self):
+    #
+    # This test implements single kernel forward and 2 kernel backward with
+    # tc.define_with_autograd. The performance of the backward is expected to be
+    # significantly better because the loops types in the single kernel may not
+    # be currently fused profitably (need to investigate the fused case deeper).
+    # The API is more bloated than hoped but it is unclear we can do better,
+    # suggestions most welcome.
+    # Maybe consider making cache_filename and tuner_config global parameters.
+    #
+    def test_conv_backward_2kernels(self):
+        conv = """
+        def convolution(float(N,C,H,W) I, float(M,C,KH,KW) W1) -> (O) {
+            O(n, m, h, w) +=!
+                I(n, r_c, h + r_kh, w + r_kw) * W1(m, r_c, r_kh, r_kw)
+        }
+        def convolution_igrad(float(M,C,KH,KW) W1, float(N,M,H,W) d_O)
+            -> (d_I)
+        {
+            d_I(n, c, h, w) +=!
+                d_O(  n, r_m, h - r_kh, w - r_kw) * W1(r_m, c, r_kh, r_kw)
+        }
+        def convolution_wgrad(float(N,C,H,W) I, float(N,M,H,W) d_O) -> (d_W1)
+        {
+            d_W1(m, c, kh, kw) +=!
+                d_O(r_n,   m, r_h - kh, r_w - kw) *  I(r_n, c,  r_h,  r_w)
+        }
+        """
+
+        N, C, H, W, O, kH, kW = 32, 4, 56, 56, 16, 1, 1
+        cache_filename = "/tmp/" + str(uuid.uuid4())
+        convolution = tc.define_with_autograd(
+            conv,
+            forward_entry_points=("convolution", ),
+            backward_entry_points=("convolution_igrad", "convolution_wgrad"),
+            cache_filename=cache_filename,
+            # First: (I, W1, d_O) -> (W1, d_O) => (1, 2)
+            # Second: (I, W1, d_O, d_I) -> (I, d_O) => (0, 2)
+            backward_input_indices=((1, 2), (0, 2), ),
+            tuner_config=tuner_config)
+        I = Variable(torch.randn(N, C, H, W).cuda(), requires_grad=True)
+        W = Parameter(torch.randn(O, C, kH, kW).cuda())
+        # First occurrence on a new cache filename triggers tuning
+        out, = convolution(I, W)
+        out.sum().backward()
+        # Subsequent occurrences do not
+        out, = convolution(I, W)
+        out.sum().backward()
+        # TODO: test results!!!
```

Comment:
You might be interested in `torch.autograd.gradcheck` (see e.g. right above [this paragraph](https://pytorch.org/docs/stable/notes/extending.html?highlight=gradcheck#extending-torch-nn)).

Reply:
There's also `gradgradcheck` if you care about higher order derivatives

Reply:
Also, since the `TODO: test results` appears in so many places, it's probably a good idea to have a common testing procedure with a signature similar to `gradcheck` (take the function and example inputs, do some checks). This way you write it once and any subsequent improvements apply to all functions you check.